tests/stub/authorization/test_renewable_auth.py

```diff
@@ -0,0 +1,300 @@
+        def _test(error_, routing_):
+            count = 0
+
+            def provider():
+                nonlocal count
+                count += 1
+                credentials = "password" if count > 1 else "pass"
+
+                return types.RenewableAuthToken(
+                    types.AuthorizationToken(
+                        scheme="basic",
+                        principal="neo4j",
+                        credentials=credentials
+                    ),
+                    10_000
+                )
+
+            expected_call_count = 2 if error_ == "token_expired" else 1
+
+            self.start_server(
+                self._reader,
+                self.script_fn_with_features(f"reader_reauth_{error_}.script")
+            )
+            if routing_:
+                self.start_server(
+                    self._writer,
+                    self.script_fn_with_features(
+                        f"writer_reauth_{error_}.script"
+                    )
+                )
+                self.start_server(self._router, "router_single_reader.script")
+
+            with self.driver(provider, routing=routing_) as driver:
+                if routing_:
+                    with self.session(driver, "w") as session_w:
+                        list(session_w.run("RETURN 1 AS n"))
+
+                with self.session(driver) as session_r1:
+                    with self.session(driver) as session_r2:
+                        with self.session(driver) as session_r3:
+                            # bind connection 1
+                            s1_tx = session_r1.begin_transaction()
+                            list(s1_tx.run("RETURN 1.1 AS n"))
+
+                            self.assertEqual(1, count)
+
+                            # bind connection 2
+                            s2_tx = session_r2.begin_transaction()
+                            list(s2_tx.run("RETURN 2.1 AS n"))
+
+                            # bind connection 3
+                            s3_tx = session_r3.begin_transaction()
+                            list(s3_tx.run("RETURN 3.1 AS n"))
+
+                            s2_tx.commit()
+
+                            self.assertEqual(1, count)
+
+                            with self.assertRaises(types.DriverError) as exc:
+                                # connection 2 fails, gets closed
+                                list(session_r2.run("RETURN 2.2 AS n"))
+                            if error_ == "token_expired":
+                                self.assert_is_token_error(exc.exception)
+                            elif error_ == "authorization_expired":
+                                self.assert_is_authorization_error(
+                                    exc.exception
+                                )
+                            else:
+                                raise ValueError(f"Unknown error {error_}")
+                            self.assertEqual(expected_call_count, count)
```

Comment:
It might be worth supporting both `1` and `2` in context of `token_expired` here. In context of Java driver it seemed to make more sense to call the authentication provider when credentials would be needed again. In this case it would be on line `246`.

Reply:
Yeah, makes sense. Feel free to move that line down to after the next `begin_transaction`.